doc:  add a generic section on exploring package sets

fricklerhandwerk opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fricklerhandwerk committed:
doc: don't recommend listing packages with `nix-env`

Co-authored-by: wamirez <wamirez@protonmail.com>

Atemu commented on doc/languages-frameworks/haskell.section.md:
````diff
@@ -25,21 +25,12 @@ All of these packages are originally defined in the `haskellPackages` package
 
-```console
-$ nix-env -f '<nixpkgs>' -qaP -A haskellPackages
-haskellPackages.a50                                                         a50-0.5
-haskellPackages.AAI                                                         AAI-0.2.0.1
-haskellPackages.aasam                                                       aasam-0.2.0.0
-haskellPackages.abacate                                                     abacate-0.0.0.0
-haskellPackages.abc-puzzle                                                  abc-puzzle-0.2.1
````
The new method does not show version information which might be critical.

It'd be great to retain the example output in any case though.

sternenseemann reviewed (dismissed):
This just seems silly to me. There's nothing wrong with using `nix-env` in this way and the presented alternative is strictly worse?

Atemu commented:
Listing packages given a (possibly nested) attrset while ignoring any eval errors is one of the only sensible uses of nix-env. What we want to get away from is using nix-env for imperative package management because it really sucks quite a lot at that.

fricklerhandwerk committed:
make a proper example

fricklerhandwerk commented:
@Atemu would I be missing something crucial if I claim this can be done in the 4-liner I added as an example?

Atemu commented:
Yes, it'd fall on its nose on the first attribute that doesn't eval or doesn't have a version. haskellPackages is quite well behaved in this regard but go I dare you to this on `pkgs`... The only way to get this sort of information reliably on `pkgs` is `nix-env`; I've been there.

fricklerhandwerk commented:
Incredible...

https://github.com/NixOS/nix/blob/master/src/nix-env/nix-env.cc#L195
https://github.com/NixOS/nix/blob/master/src/libexpr/get-drvs.cc#L311-L330

How about we all forget that code ever existed, merge this PR, and focus on something else?

Atemu commented:
I don't have a strong opinion on this and I don't maintain this part of Nixpkgs. I also don't have anything more to contribute to this discussion, so I'm going to unsubscribe here.

fricklerhandwerk commented:
@maralorn @ncfavier opinions?

maralorn commented:
I have no strong opinion on this and I frankly don’t feel like arbitrating this on my free time. So I will just leave some thoughts:

I empathize with the desire to get rid of references to nix-env. I mean in general even mentioning that packages won’t get listed by `nix-env -qa` seems quite irrelevant because that workflow seems rather alien to me. I wonder if anyone ever actually needs this on the command line.

On the other hand the fact that this command becomes so much more complicated without it is annoying, eval errors will slip in from time to time and this usage of nix-env seems harmless in contrast to the usual critique of that command.

 I think the biggest advantage of having the listing in the manual is that it makes the package set a bit more graspable.

My usual workflow is search.nixos.org or nix repl + tabcompletion, I don‘t know if that’s representative but if it is it might be the way to go also for the manual …

sternenseemann commented:
> Incredible... just ignore assertions?

You fundamentally have to do this when dealing with how nixpkgs works. `haskellPackages` is a *very* tame corner of nixpkgs, but even here you'll need to deal with assertions if you want to know `outPath`s.

> the only stable, reliable, and (mostly) sane interface to Nix is the Nix language.

It's frankly bizarre to state that to list packages you should

1. Install `jq`,
2. Copy and paste some code snippet that is specific to `haskellPackages`

instead of using a subcommand of `nix-env(1)` that is not problematic.

> I thought it was widely agreed-upon that using `nix-env` should not be encouraged

Yes, `nix-env` for package management is to be discouraged, but that concerns `nix-env -u` and `nix-env -i`. I just don't see any alternative to `nix-env -qP` in this case.

fricklerhandwerk committed:
add a generic section on exploring package sets

this removes the hard-coded listing, which can later be replaced by
a dynamic one as for the Python interpreters

fricklerhandwerk committed:
fix broken reference

fricklerhandwerk committed:
shorter repl example

maralorn commented:
I like this in general.

I still think that having a few example packages listed in the Haskell section makes it easier to quickly grasp the structure of the package sets in question. But overall this change is already an improvement.

fricklerhandwerk committed:
clarify why using `nix-env --query` makes sense
